cmd/zfs/zfs_main.c

```diff
@@ -1161,13 +1173,12 @@ destroy_callback(zfs_handle_t *zhp, void *data)
 	 * because we must delete a clone before its origin.
 	 */
 	if (zfs_get_type(zhp) == ZFS_TYPE_SNAPSHOT) {
+		cb->cb_snap_count++;
 		fnvlist_add_boolean(cb->cb_batchedsnaps, name);
+		if (cb->cb_snap_count % 10 == 0)
```

Comment:
One question, did you do any experimentation to settle on 10 for the optimal batch size?  It sounds reasonable, but my concern is that is this value is too small it will end up forcing additional TXG syncs which could end up negatively impacting performance.

Reply:
Hi Brian.  I spent some time looking into this question, sticking with the batch size of 10.  First I built up some sample data to have some snaps with some blocks in their deadlists.

```
#!/bin/ksh -p

#
#
#
for x in {1..50} ; do
	cp /tank/file /tank/fs1/file$x
	(( i = x - 1))
	[[ $i > 1 ]] && rm /tank/fs1/file$i
	zfs snapshot tank/fs1@snap$x
done
```
Since I've arranged to have the code triggered by -d, I issued the following:
```
zfs destroy -d tank/fs1@%
```
Because of async destroy, all 50 snaps got queued up for destroy in the same txg:
```
520627896   spa_history.c:317:spa_history_log_sync(): txg 28114 destroy tank/fs1@snap12 (id 604) 
1520627896   spa_history.c:317:spa_history_log_sync(): txg 28114 destroy tank/fs1@snap14 (id 610) 
1520627896   spa_history.c:317:spa_history_log_sync(): txg 28114 destroy tank/fs1@snap38 (id 682) 
1520627896   spa_history.c:317:spa_history_log_sync(): txg 28114 destroy tank/fs1@snap26 (id 646) 
...
...
```
And with the block delete activity that got triggered, the txg time seemed to stick with the normal 5 seconds.
```
1520627918   txg.c:543:txg_sync_thread(): txg=28119 quiesce_txg=28120 sync_txg=28114
1520627918   txg.c:510:txg_sync_thread(): waiting; tx_synced=28119 waiting=28114 dp=ffff8801ea948000
1520627923   txg.c:543:txg_sync_thread(): txg=28120 quiesce_txg=28121 sync_txg=28114
1520627923   txg.c:510:txg_sync_thread(): waiting; tx_synced=28120 waiting=28114 dp=ffff8801ea948000
1520627928   txg.c:543:txg_sync_thread(): txg=28121 quiesce_txg=28122 sync_txg=28114
1520627928   txg.c:510:txg_sync_thread(): waiting; tx_synced=28121 waiting=28114 dp=ffff8801ea948000
1520627933   txg.c:543:txg_sync_thread(): txg=28122 quiesce_txg=28123 sync_txg=28114
1520627933   txg.c:510:txg_sync_thread(): waiting; tx_synced=28122 waiting=28114 dp=ffff8801ea948000
1520627938   txg.c:543:txg_sync_thread(): txg=28123 quiesce_txg=28124 sync_txg=28114
1520627938   txg.c:510:txg_sync_thread(): waiting; tx_synced=28123 waiting=28114 dp=ffff8801ea948000
1520627943   txg.c:543:txg_sync_thread(): txg=28124 quiesce_txg=28125 sync_txg=28114
````
Does this look ok to you?

Reply:
Yup, that all looks reasonable.  Thanks for looking in to this.